Verified both Henyey-Greenstein scattering and refractive index mismatch

the-professor510 opened the pull request:
Fixed Henyey-Greenstein scattering and verified it with [MCML](https://doi.org/10.1016/0169-2607(95)01640-F). This can be repeated using validation1.toml as the input file. Both validation2.toml and validatoin3.toml are used to verify the refractive index mismatch against [this](https://doi.org/10.1117/12.146307), this is a repeat of the original validation used. Added functionality to multiple source types, have added the moss egg geometry, and added more functionality to the sphere and box geometries. Verified the circle detector, and added an emission location output.

lewisfish commented on app/main.f90:
```diff
@@ -1,6 +1,6 @@
 program mcpolar
 !! Entry point for program
-    use kernels, only : weight_scatter, pathlength_scatter
+    use kernels, only : weight_scatter, pathlength_scatter, pathlength_scatter2
```
probably needs a better name...

lewisfish commented on src/detectors/detectors.f90:
```diff
@@ -22,6 +22,39 @@ module detectors
+        real(kind=wp) :: f2Aperture
+        !> distance between the detector plane and the front lens
+        real(kind=wp) :: frontOffset
+        !> distance between the fibre plance and the back lens
```
plane not plance

lewisfish commented on src/iarray.f90:
what does first shot mean here? also remove unused code please

lewisfish commented on src/sdfs/sdf_base.f90:
remove unused code

lewisfish commented on src/setup.f90:
```diff
@@ -29,6 +29,8 @@ subroutine setup_simulation(sdfarray, dict)
             call alloc_array(settings%grid%nxg, settings%grid%nyg, settings%grid%nzg)
             call zarray()
 
+            !print*, settings%experiment
```
remove unused code

lewisfish commented on src/writer.f90:
```diff
@@ -45,35 +45,76 @@ subroutine normalise_fluence(grid, array, nphotons)
             do i = 1, size(dects)
-                open(newunit=u, file=trim(fileplace)//"detectors/detector_"//str(i)//".dat")
+                !open(newunit=u,  access="stream",&
+                !     status='REPLACE', file=trim(fileplace)//"detectors/detector_"//str(i)//".dat")
```
remove unused code

lewisfish commented on src/parse/parse_geometry.f90:
```diff
@@ -0,0 +1,302 @@
+            call get_value(child, "boundingBox", children, requested=.false., origin=origin)
+            if(associated(children))then
+                nlen = len(children)
+                if(nlen .NE. 3)then
```
Same here, don't use .ne. use /=

lewisfish commented on res/scat_test.toml:
scat_test.toml is used in a test, so the original values need to be restored in order for the tests to run successfully

lewisfish commented on tools/plotDetectors.py:
```diff
@@ -0,0 +1,93 @@
+    #print(count)        
+    totalCounts = sum(count)
+    print("Detector Type : " + dectType)
+    print("Total Diffuse : " + str(totalCounts/nPackets))
```
Use f-strings instead of str(foo) https://realpython.com/python-f-strings/#doing-string-interpolation-with-f-strings-in-python

lewisfish commented on tools/validateFibreDect.py:
use f-strings

lewisfish commented on tools/validateHGG.py:
use f-strings

lewisfish commented on tools/validateRIMismatch.py:
should probably provide a common read nrrd functionality for all plotting files, instead of copying it over into each file

lewisfish commented:
Hey @the-professor510, This is getting too big for me to go through. If you stop adding to this PR and I'll get though it in the next month or so. After that I'll add you as a maintainer so you can directly commit to the repo so there is no backlog.
You should be on the same page as me style wise for the repo after I've reviewed it all anyways

the-professor510 commented:
Hi @lewisfish I have just made one final change in the code to ensure that all unit tests were passed successfully. I will branch my current fork to stop adding to this PR while I work on adding an inverse Monte Carlo functionality. 

To give a run down of the significant additions I have added multiple compilation flags for different key functionalities. These functionalities are as follows:

- Normal Forward MCRT (no extra compile time flag)
- Calculating Escape Function ( --flag "-DescapeFunction")
- INCOMPLETE: Inverse MCRT (--flag "-DinverseMCRT")

Normal forward MCRT does as describes and runs MCRT forwards with the user defined source, geometry, and detectors. It will output the data from the detectors, an emission grid (where packets were launched from), and an absorption grid (calculated by dropping total weight) by default.

Calculating the Escape Function will give a grid output related to each detector, this is the probability of a Raman or fluorescence photon being detected from every location in the grid by each detector. This method works by creating an isotropic source at each location in a user defined Cylindrical or Cartesian grid and calculating the efficiency of the light reaching each detector for each source location. This will take a long time to run, so I have added different symmetries for Cylindrical and Cartesian grids to reduce the time to run. Note, the element-wise multiplication of a Normal Forward MCRT absorption grid and an escape function grid will return the Raman detection efficiency for a given illumination and detector geometry. 

The aim of Inverse MCRT is to be able to estimate the optical properties of mus, mua, hgg, and n when given a target value for the total intensity on multiple detectors (more detectors than optical properties being estimated). This is a work in progress and has not been tested. The current plan includes the implementation of AdaLIPO to find the global minimum with minimal evaluations of our cost function (running MCRT) and no knowledge of the maximum gradient.  

There are two other compile time flags that change the functionality of the program:
- calculate and record the fluence by the pathlength method (--flag "-Dpathlength")
- use survival biasing for variance reduction (--flag "-DsurvivalBias")

Calculating and recording the fluence can only be used for normal forward MCRT, while survival biasing for variance reduction can be used with all other compile flags.
